src/rgw/rgw_data_sync.cc

```diff
@@ -1774,20 +1806,70 @@ class RGWDataSyncShardCR : public RGWCoroutine {
     }
     return 0;
   }
+};
 
-  utime_t get_idle_interval() const {
-    ceph::timespan interval = std::chrono::seconds(cct->_conf->rgw_data_sync_poll_interval);
-    if (!ceph::coarse_real_clock::is_zero(error_retry_time)) {
-      auto now = ceph::coarse_real_clock::now();
-      if (error_retry_time > now) {
-        auto d = error_retry_time - now;
-        if (interval > d) {
-          interval = d;
-        }
+class RGWDataSyncShardCR : public RGWCoroutine {
+  RGWDataSyncCtx* const sc;
+  const rgw_pool& pool;
+  const uint32_t shard_id;
+  rgw_data_sync_marker& sync_marker;
+  RGWSyncTraceNodeRef& tn;
+
+  ceph::mutex inc_lock = ceph::make_mutex("RGWDataSyncShardCR::inc_lock");
+  bc::flat_set<rgw_data_notify_entry> modified_shards;
+
+public:
+  RGWDataSyncShardCR(RGWDataSyncCtx* const _sc, const rgw_pool& pool,
+                     const uint32_t shard_id,
+		     rgw_data_sync_marker& sync_marker,
+                     RGWSyncTraceNodeRef& tn,
+		     bool* /* XXX We were passing this in as reset_backoff but we
+			      never did anything with it. If we should be doing
+			      something with it, we should put that in. If not
+			      we should just remove it from the supplied
+			      parameters. */)
```

Comment:
DataSyncShardCR is wrapped in a BackoffControlCR, which will keep retrying DataSyncShardCR after it returns an error

if we have two rgws running in the same zone, we'd expect one of them to grab the lock on this shard and process it, while the other would keep looping on `EBUSY`

so the backoff behavior is mostly controlling how we poll this lock - we want to keep backing off until we succeed in acquiring it

if we're moving the management of `RGWContinuousLeaseCR` into `RGWDataSyncShardCR` as i suggested above, that should make it easier to manage this flag. but you're also welcome to leave this as a TODO comment

Reply:
Stuck a TODO on the flag.